### DISCUSSION TITLE
Using multi-arch Docker images with Pulumi?

### DISCUSSION THREAD

**Discussion by aterreno** [pulumi/pulumi#7428] [Pulumi Providers]
Hi all,
I think I've hit a bug with pulumi building docker images, I am trying to build a multiarch image, in order to do that I've setup my mac (M1) to use buildx by default.
When I add the --platform flag into the pulumi code I get back an error, that the command is not right.
If I copy the command and paste into the terminal as-is the docker build works...
I've put a very simple sample into github to reproduce https://github.com/aterreno/pulumi-docker-multiarch

What's the best repo to open an issue on github @pulumi/pulumi?

Or am I missing something else and it's not a bug?

Thanks

**Comment by mattstratton** [comment]
Tracking this issue in https://github.com/pulumi/pulumi-docker/issues/296

Once this is resolved, we should update this post to help people in the future!

**Reply by ffMathy** [reply]
This is now resolved.

**Comment by mattstratton** [comment]
(I updated the title of this post and moved it to the providers category so that once we have the answer, it will work properly here! Thanks for posting it!!)

**Reply by spara** [reply]
I've hit this issue before:

```sh
docker build --push --platform  linux/arm64,linux/arm/v7 -t my-repo ./app
```

```python
image = docker.Image('my-image',
    build=docker.DockerBuild(context=f'./{my_image}',
        extra_options={'--push', '--platform', 'linux/arm64,linux/arm/v7'}),
    image_name=image_name,
    registry=registry_info,
)
```

In @aterreno's example, passing in the whole string doesn't work properly, see https://github.com/pulumi/pulumi-docker/pull/260

You also have to add `--push` flag for buildx

However, my example is unreliable because the string for extra_options isn't read in order. Note that buildx is enabled.

**Reply by aterreno** [reply]
I see @spara 

I am trying this now

```typescript
import * as awsx from "@pulumi/awsx";

const repository = new awsx.ecr.Repository("test");

const image = repository.buildAndPushImage({
  context: ".",
  dockerfile: "./Dockerfile",
  extraOptions: ["--platform", "linux/arm64", "--platform", "linux/amd64"],
});

module.exports = { image, repository };
```

**Reply by aterreno** [reply]
but the last --platform wins over the other one. If I put them with a comma it doesn't work. 
And buildx doesn't like --push btw...

**Reply by aterreno** [reply]
really strange, all the code is doing is just pushing args

https://github.com/pulumi/pulumi-docker/blob/master/sdk/nodejs/docker.ts#L466

I don't see why that would work 🤔

**Reply by aterreno** [reply]
debugging further, if I log the command line params looks like:

```js
["build","-f","./Dockerfile","--quiet","--platform","linux/amd64,linux/arm64,linux/arm/v7"]
```

**Reply by aterreno** [reply]
later on we endup doing a call to child_process.spawn with:

```js    
{"cmd":"docker","args":["build","-f","./Dockerfile","--quiet","--platform","linux/amd64,linux/arm64,linux/arm/v7",".","-t","e23d430b-container"]}
```

**Reply by aterreno** [reply]
I think I've got it, the child process doesn't understand that we are doing a `docker buildx create --use` before

This code works for example 

```js
const { spawn } = require("child_process");

const command = {
    cmd: 'docker',
    args: [
      'buildx',
      'build',
      '--platform',
      'linux/amd64,linux/arm64,linux/arm/v7',
      '.',
    ]
  }

const p = spawn(command.cmd, command.args, process.env);

p.stdout.on("data", (data) => {
  console.log(`stdout: ${data}`);
});

p.stderr.on("data", (data) => {
  console.error(`stderr: ${data}`);
});

p.on("close", (code) => {
  console.log(`child process exited with code ${code}`);
});
```

**Reply by aterreno** [reply]
so, remarkably probalby this works

```ts
import * as awsx from "@pulumi/awsx";

const repository = new awsx.ecr.Repository("test");

const image = repository.buildAndPushImage({  
  extraOptions: [
    `buildx`,
    `build`,
    '--platform',
    'linux/amd64,linux/arm64,linux/arm/v7',
  ],
});

module.exports = { image, repository };

```

but indeed, once you start specifying other params such as -f and so on the order is wrong, so it's not a viable workaround for most use cases.

**Comment by nimbinatus** [comment]
Adding a workaround from @stevesloka while the issue is still open:
https://github.com/pulumi/pulumi-docker/issues/296#issuecomment-1030094518
> I found a quick workaround for anyone needing to build linux/amd64 images on their m1-based macs. If you set the env variable `DOCKER_DEFAULT_PLATFORM=linux/amd64` before your `docker build` (or any command), docker for mac will build the proper image for you.
>
>This doesn't take the place of proper multi-arch builds, but if this can unlock anyone on m1's for now I wanted to share.
